### PR TITLE
Support HTTP headers on HttpWaitStrategy

### PR DESCRIPTION
**Pull request by renatomefi**
This enables webservers that depend on headers for healthchecks to be able to correctly be queried by the HttpWaitStrategy.

As there are no unit tests for the HttpWaitStrategy I'm using the container logs as confirmation of the correct HTTP call execution.

### CODE REVIEW

File: core/src/test/java/org/testcontainers/junit/wait/strategy/AbstractWaitStrategyTest.java [hunk 1]
```diff
@@ -118,4 +118,9 @@ protected void waitUntilReadyAndSucceed(GenericContainer container) {
         assertTrue(String.format("Expected container to be ready after timeout of %sms",
             WAIT_TIMEOUT_MILLIS), ready.get());
     }
+
+    protected String waitUntilReadyAndSucceedWithLogs(GenericContainer container) {
```
Comment (kiview) [inline core/src/test/java/org/testcontainers/junit/wait/strategy/AbstractWaitStrategyTest.java:122]: I am not that happy about adding this method.

File: core/src/test/java/org/testcontainers/junit/wait/strategy/HttpWaitStrategyTest.java [hunk 2]
```diff
@@ -30,6 +34,26 @@ public void testWaitUntilReadyWithSuccess() {
...
+        HashMap<String, String> headers = new HashMap<>();
+        headers.put("baz", "boo");
+        String logs = waitUntilReadyAndSucceedWithLogs(
+            startContainerWithCommand(createShellCommand("200 OK", GOOD_RESPONSE_BODY),
+                createHttpWaitStrategy(ready)
+                    .withHeader("foo", "bar")
+                    .withHeaders(headers)
+            ));
+
+        assertThat(logs, containsString("foo: bar"));
```
Comment (kiview) [inline core/src/test/java/org/testcontainers/junit/wait/strategy/HttpWaitStrategyTest.java:53]: I would much prefer a black box test with a server requiring the headers.
On the other hand, if you want logs, you can just access the container, which is returned by `startContainerWithCommand`.

File: core/src/test/java/org/testcontainers/junit/wait/strategy/AbstractWaitStrategyTest.java (hunk 1 above)
Comment (renatomefi) [reply core/src/test/java/org/testcontainers/junit/wait/strategy/AbstractWaitStrategyTest.java:122]: If we proceed with checking logs, should I return the container itself?

File: core/src/test/java/org/testcontainers/junit/wait/strategy/HttpWaitStrategyTest.java (hunk 2 above)
Comment (renatomefi) [reply core/src/test/java/org/testcontainers/junit/wait/strategy/HttpWaitStrategyTest.java:53]: I got a bit blocked by how the tests works, the netcat utility is quite limited, I was afraid that I was going to have to touch too many things in order to achieve the result needed!
If you think that I should follow this path, what kind of service should I spin off? A mock service, custom bash script? what'd be appropriate in your opinion?

File: core/src/test/java/org/testcontainers/junit/wait/strategy/AbstractWaitStrategyTest.java (hunk 1 above)
Comment (kiview) [reply core/src/test/java/org/testcontainers/junit/wait/strategy/AbstractWaitStrategyTest.java:122]: Method in general not necessary.

File: core/src/test/java/org/testcontainers/junit/wait/strategy/HttpWaitStrategyTest.java (hunk 2 above)
Comment (kiview) [reply core/src/test/java/org/testcontainers/junit/wait/strategy/HttpWaitStrategyTest.java:53]: After thinking a bit more about it, I think your approach is actually valid and does not involve too much overhead.

I rewrote the method like this:

```java
    @Test
    public void testWaitUntilReadyWithSuccessWithCustomHeaders() {
        HashMap<String, String> headers = new HashMap<>();
        headers.put("baz", "boo");
        GenericContainer container = startContainerWithCommand(createShellCommand("200 OK", GOOD_RESPONSE_BODY),
            createHttpWaitStrategy(ready)
                .withHeader("foo", "bar")
                .withHeaders(headers)
        );
        waitUntilReadyAndSucceed(container);

        String logs = container.getLogs();

        assertThat(logs, containsString("foo: bar"));
        assertThat(logs, containsString("baz: boo"));
    }
```

Comment (renatomefi) [reply core/src/test/java/org/testcontainers/junit/wait/strategy/HttpWaitStrategyTest.java:56]: Thanks, I've updated it with your suggestion!

**Comment by rnorth** [comment]
Build failure - I'll merge from master to ensure that we're using the latest, stable (😄) master build as a basis.

**Comment by renatomefi** [comment]
Hey @rnorth and @kiview, I've rebased it with master today, and it's looking good! :)
Are you still considering merging this one?

thanks

**Comment by renatomefi** [comment]
I think it's ready to merge?

**Review by rnorth — APPROVED** [review]

**Pull request merged**